File: C++/Math/9. Palindrome Number.cpp
// INT_MAX = 2^31 - 1 = 2147483647, INT_MAX = 2^32 = 2147483648

class Solution {
public:
    bool isPalindrome(int x) {
        const int CAP = INT_MAX/10;
        if (x < 0) return false;
        int backup = x;
        int reverse = 0;
        while (x) {
            if (reverse > CAP || (reverse == CAP && x%10>=8)) return false;
            reverse = reverse * 10 + x%10;
            x /= 10;
        }
        return reverse == backup;
    }
};

File: Python/Archive/9. Palindrome Number.py
# INT_MAX = 2147483647

class Solution:
    def isPalindrome(self, x: int) -> bool:
        CAP = 214748364
        if x < 0: return False
        backup = x
        reverse = 0
        while x:
            if reverse > CAP or (reverse == CAP and x%10 >= 8): return False
            reverse = reverse * 10 + x%10
            x //= 10
        return reverse == backup